src/peft/peft_model.py

```diff
@@ -918,6 +919,22 @@ def forward(
             return self.base_model(
                 input_ids=input_ids, decoder_input_ids=decoder_input_ids, past_key_values=past_key_values, **kwargs
             )
+        elif peft_config.peft_type == PeftType.PROMPT_TUNING:
+            if inputs_embeds is None:
+                inputs_embeds = self.word_embeddings(input_ids)
+
+            if attention_mask is not None:
+                # concat prompt attention mask
+                prefix_attention_mask = torch.ones(batch_size, peft_config.num_virtual_tokens).to(
+                    attention_mask.device
+                )
+                kwargs["attention_mask"] = torch.cat((prefix_attention_mask, attention_mask), dim=1)
+
+            prompts = self.get_prompt(batch_size=batch_size)
+            prompts = prompts.to(inputs_embeds.dtype)
+            inputs_embeds = torch.cat((prompts[:, : peft_config.num_virtual_tokens], inputs_embeds), dim=1)
```

Comment:
I went over the papers again. Prefix-Tuning paper is adding prefixes to both the encoder and decoder while prompt-tuning paper seems only to add them to the input. 

Prompt Tuning Paper:

> Given a series of n tokens, {x1, x2, . . . , xn}, the
> first thing T5 does is embed the tokens, forming
> a matrix Xe ∈ R
> n×e where e is the dimension of
> the embedding space. Our soft-prompts are represented as a parameter Pe ∈ R
> p×e
> , where p is the
> length of the prompt. Our prompt is then concatenated to the embedded input forming a single matrix [Pe; Xe] ∈ R
> (p+n)×e which then flows though
> the encoder-decoder as normal. Our models are
> trained to maximize the probability of Y , but only
> the prompt parameters Pe are updated.

Prefix-Tuning paper
<img width="1283" alt="Screenshot 2023-06-16 at 1 45 59 PM" src="https://github.com/huggingface/peft/assets/13534540/8e69d33c-fd6d-483a-bcc8-e146be3afc62">

Therefore, you are correct. Thank you!

Reply:
Could you please then remove the need for prompt tokens in the decoder for P-Tuning approach too. This will lead to both of these being supported by `generate` for `seq2seq` tasks.

Also, remove the point 3 from the caveats section of README.md as this PR solves it.

> 3. For encoder-decoder models, P_TUNING or PROMPT_TUNING doesn't support generate functionality of transformers because generate strictly requires decoder_input_ids but P_TUNING/PROMPT_TUNING appends soft prompt embeddings to input_embeds to create new input_embeds to be given to the model. Therefore, generate doesn't support this yet.

Reply:
Awesome! I'll be working on it on Monday 😉

Reply:
@thomas-schillaci @pacman100 Thanks for your valuable contribution. What is about the point 2, "When using P_TUNING or PROMPT_TUNING with SEQ_2_SEQ task, remember to remove the num_virtual_token virtual prompt predictions from the left side of the model outputs during evaluations."? Could you please also update it?